lib/fluent/plugin/filter_kubernetes_metadata.rb

```diff
@@ -163,55 +172,80 @@ def configure(conf)
+
+        this     = self
+        metadata = @cache.getset(cache_key) {
+          if metadata
+            kubernetes_metadata = this.get_metadata(
+              metadata['kubernetes']['namespace_name'],
+              metadata['kubernetes']['pod_name'],
+              metadata['kubernetes']['container_name']
+            )
+            metadata['kubernetes'] = kubernetes_metadata if kubernetes_metadata
+            metadata
+          end
+        }
+        if pod_id && (pod_id != metadata['kubernetes']['pod_id'])
+          log.debug("pod_id #{match_data['pod_id']} from log not equal to pod_id #{metadata['kubernetes']['pod_id']} from kubernetes for #{cache_key}")
+        end
+        if @include_namespace_id
+          namespace_name = metadata['kubernetes']['namespace_name']
+          namespace_id = @namespace_cache.getset(namespace_name) {
+            namespace = @client.get_namespace(namespace_name)
+            namespace['metadata']['uid'] if namespace
+          }
+          metadata['kubernetes']['namespace_id'] = namespace_id if namespace_id
+        end
+      end
+      metadata
+    end
+
     # NOTE: fluentd requires that records/hashes have string keys, not symbol keys
     # http://docs.fluentd.org/articles/plugin-development#record-format
     def filter_stream(tag, es)
       new_es = MultiEventStream.new
 
-      match_data = tag.match(@tag_to_kubernetes_name_regexp_compiled)
-
-      if match_data
-        metadata = {
-            'docker' => {
-                'container_id' => match_data['docker_id']
-            },
-            'kubernetes' => {
-                'namespace_name' => match_data['namespace'],
-                'pod_name'       => match_data['pod_name'],
-                'container_name' => match_data['container_name']
-            }
-        }
-
-        if @kubernetes_url.present?
-          cache_key = "#{metadata['kubernetes']['namespace_name']}_#{metadata['kubernetes']['pod_name']}_#{metadata['kubernetes']['container_name']}"
-
-          this     = self
-          metadata = @cache.getset(cache_key) {
-            if metadata
-              kubernetes_metadata = this.get_metadata(
-                metadata['kubernetes']['namespace_name'],
-                metadata['kubernetes']['pod_name'],
-                metadata['kubernetes']['container_name']
-              )
-              metadata['kubernetes'] = kubernetes_metadata if kubernetes_metadata
-              metadata
-            end
-          }
-          if @include_namespace_id
-            namespace_name = metadata['kubernetes']['namespace_name']
-            namespace_id = @namespace_cache.getset(namespace_name) {
-              namespace = @client.get_namespace(namespace_name)
-              namespace['metadata']['uid'] if namespace
-            }
-            metadata['kubernetes']['namespace_id'] = namespace_id if namespace_id
-          end
+      unless @use_journal
```

Comment:
The logic for journald is pretty different to that for straight JSON docker logs because it doesn't use the fluentd tag for original container info so how about splitting this into separate functions for journald/json logging, determining which to use in `configure` based on the `use_journal` config option & then calling that in `filter_stream`? Gets away from the need for `unless @use_journal`/`if @use_journal`/etc which should give a bit of (probably unnecessary but nice to have) perf boost as well as being easier to reason about flows.